File: matlab/tests/test_routing_parity_round3.m
%% test_routing_parity_round3.m — regressions from the routing-parity audit, round 3
%
%  Mirror of the Python tests/test_routing_parity_round3.py. The round
%  removed the dead code the audit found and made explainDispatch
%  report the route the call takes; each block pins one of those
%  outcomes so the two languages keep agreeing:
%
%    explainDispatch on a flat cosine builds the selector's inputs by
%      the same function the call uses (INTERNAL.FLATSELECTORINPUTS: the
%      wrap vector, the per-attribute grid node counts, the memo flags),
%      applies the empty-operand rule before the selector and the
%      ordered-attribute rule after it --- so on a rel-per pair above
%      the sigma/P threshold the report follows the declared wrap as the
%      call does, and on an ordered pair it names Bulger's method under
%      any method;
%    D-6   the eval cost model prices the spectral branch of the Möbius
%      relative evaluator wherever that branch engages (no mode-grid
%      decline, since the evaluator has none);
%    B-16  the Möbius per-attribute matrix has one abs r >= 2 route,
%      and agrees with the enumerated reference at K = r;
%    B-16  the tuple-centres closed form refuses an inner [rel] unit;
%    D     the deleted names stay deleted.
%
%  Standalone-runnable; appends to `results` when called from test_mpt.m.
%  mptTestIsolateDefaults sets truncationSigmas = Inf (accuracy floor).

if ~exist('results', 'var')
    results = {};
    standalone = true;
    addpath(fileparts(mfilename('fullpath')));
    addpath(fullfile(fileparts(mfilename('fullpath')), 'reference'));
    addpath(fileparts(fileparts(mfilename('fullpath'))));
    clear cleanupDefaults_rp3
    cleanupDefaults_rp3 = mptTestIsolateDefaults(); %#ok<NASGU>
else
    standalone = false;
end

rp3_P = 12.0;
rp3_absWarn = warning('off', 'buildExpTens:absPerSingleImage');
rp3_relWarn = warning('off', 'buildExpTens:isRelDegenerate');
rp3_warnCleanup = onCleanup(@() cellfun(@warning, ...
    {rp3_absWarn, rp3_relWarn})); %#ok<NASGU>

% --- explainDispatch follows the wrap rule above the threshold ---
% r = 3, K = 4 vs 5, rel-per at sigma/P = 0.5: above the threshold the
% declared wrap decides without pricing. The report used to omit the
% wrap vector and so named the cost model's pick. The route the call
% took is read off the memo key it wrote on the second operand.
rp3_wraps = {'full-image', 'single-image'};
rp3_expect = {'mobius', 'bulger'};
for rp3_i = 1:2
    rp3_wrap = rp3_wraps{rp3_i};
    x = rp3Flat(1, 0.5 * rp3_P, 3, true, true, rp3_wrap, rp3_P, 4);
    y = rp3Flat(2, 0.5 * rp3_P, 3, true, true, rp3_wrap, rp3_P, 5);
    rep = explainDispatch(x, y);
    [~, ~, yOut] = cosSimExpTens(x, y, 'verbose', false);
    results{end+1, 1} = sprintf( ...
        'parity round3: explain follows the %s wrap rule above the threshold', ...
        rp3_wrap); %#ok<*SAGROW>
    results{end, 2} = strcmp(rep.chosen, rp3_expect{rp3_i}) ...
        && strcmp(rep.decidedBy, 'structural rule') ...
        && rp3RouteOf(yOut, rp3_expect{rp3_i});
    results{end+1, 1} = sprintf( ...
        'parity round3: explain measure line follows the %s wrap', rp3_wrap);
    results{end, 2} = contains(rep.measure, 'transposition average') ...
        == strcmp(rp3_expect{rp3_i}, 'mobius');
end

% --- explainDispatch applies the ordered-attribute rule ---
% An ordered ([sym]=0) attribute at r > 1 has no orbit: the call takes
% Bulger's method whatever method asked for, and so does the report.
rp3_methods = {'auto', 'mobius', 'centres'};
for rp3_i = 1:3
    rp3_m = rp3_methods{rp3_i};
    rng(3, 'twister');
    x = buildExpTens({sort(rp3_P * rand(5, 2), 1)}, {[]}, 1.0, 2, ...
                     false, false, 0, false, 'verbose', false);
    rng(4, 'twister');
    y = buildExpTens({sort(rp3_P * rand(6, 2), 1)}, {[]}, 1.0, 2, ...
                     false, false, 0, false, 'verbose', false);
    rep = explainDispatch(x, y, 'method', rp3_m);
    [~, ~, yOut] = cosSimExpTens(x, y, 'method', rp3_m, 'verbose', false);
    results{end+1, 1} = sprintf( ...
        'parity round3: explain applies the ordered rule under method=%s', rp3_m);
    results{end, 2} = strcmp(rep.chosen, 'bulger') ...
        && contains(rep.decidedBy, 'ordered') ...
        && rp3RouteOf(yOut, 'bulger');
end

% --- explainDispatch uses the memo flags the call uses ---
% After a call has memoised both self inner products, the report prices
% the cross matrix alone, as the call does: its Bulger price drops.
x = rp3Flat(5, 1.0, 3, true, false, 'full-image', rp3_P, 6, 3);
y = rp3Flat(6, 1.0, 3, true, false, 'full-image', rp3_P, 6, 3);
repCold = explainDispatch(x, y);
[~, xOut, yOut] = cosSimExpTens(x, y, 'method', 'bulger', 'verbose', false);
repWarm = explainDispatch(xOut, yOut);
results{end+1, 1} = 'parity round3: explain prices the memoised self products as free';
results{end, 2} = repWarm.routeMs(1) < repCold.routeMs(1);

% --- explainDispatch applies the empty-operand rule ---
rng(7, 'twister');
pZ = sort(rp3_P * rand(5, 2), 1);
z = buildExpTens({pZ}, {zeros(5, 2)}, 1.0, 2, false, false, 0, ...
                 'verbose', false);
y = rp3Flat(8, 1.0, 2, false, false, 'full-image', rp3_P);
rep = explainDispatch(z, y);
results{end+1, 1} = 'parity round3: explain applies the empty-operand rule';
results{end, 2} = isempty(rep.chosen) ...
    && contains(rep.decidedBy, 'empty operand') ...
    && cosSimExpTens(z, y, 'verbose', false) == 0;

% --- the selector inputs are shared with the call ---
x = rp3Flat(9, 0.5 * rp3_P, 3, true, true, 'single-image', rp3_P, 4);
y = rp3Flat(10, 0.5 * rp3_P, 3, true, true, 'single-image', rp3_P, 5);
[selIn, orderedAny, nestedAny] = internal.flatSelectorInputs( ...
    internal.prunedExpTens(x), internal.prunedExpTens(y), 'cosine', []);
[~, xOut, yOut] = cosSimExpTens(x, y, 'verbose', false);
selIn2 = internal.flatSelectorInputs( ...
    internal.prunedExpTens(xOut), internal.prunedExpTens(yOut), 'cosine', []);
results{end+1, 1} = 'parity round3: flatSelectorInputs carries the wrap, per, node and memo inputs';
results{end, 2} = isequal(selIn.wrapVec, {'single-image'}) ...
    && isequal(selIn.perVec, true) && selIn.nuVec(1) > 1 ...
    && ~orderedAny && ~nestedAny ...
    && ~selIn.skipXX && ~selIn.skipYY ...
    && selIn2.skipXX && selIn2.skipYY;

% --- D-6: the eval cost model has no mode-grid decline ---
% A periodic r = 4 relative attribute at small sigma/P passes the
% evaluator's spectral gate; the model prices that branch, whose r = 4
% periodic K term is zero, so the price is the same at K = 16 and 48
% (the node path it used to fall back to carries a K-proportional
% tabulation term).
rp3_sig = 0.004 * rp3_P;
dSmall = rp3Flat(11, rp3_sig, 4, true, true, 'full-image', rp3_P, 16, 1);
dLarge = rp3Flat(12, rp3_sig, 4, true, true, 'full-image', rp3_P, 48, 1);
[~, mSmall] = internal.maEvalCostsMs(dSmall, 256);
[~, mLarge] = internal.maEvalCostsMs(dLarge, 256);
results{end+1, 1} = 'parity round3: eval cost model prices the spectral branch at r=4 periodic';
results{end, 2} = isfinite(mSmall) && isfinite(mLarge) ...
    && abs(mLarge - mSmall) <= 1e-12 * max(abs(mSmall), 1);

% --- B-16: one abs r >= 2 route agrees with enumeration at K = r ---
rng(13, 'twister');
Px = sort(40 * rand(3, 4), 1); Wx = ones(3, 4);
Py = sort(40 * rand(3, 5), 1); Wy = ones(3, 5);
rp3_prevEps = internal.accuracyFloor('setEps', 1e-300);
I = mobius.maPerAttrInnerMatrix(Px, Wx, Py, Wy, 3.0, 3, false, false, 0, ...
                                'truncationSigmas', Inf);
internal.accuracyFloor('setEps', rp3_prevEps);
Iref = zeros(4, 5);
for ii = 1:4
    for jj = 1:5
        Iref(ii, jj) = reference.innerProductDirectAbsSingleMultiset( ...
            Px(:, ii), Wx(:, ii), Py(:, jj), Wy(:, jj), 3.0, 3, false, 0);
    end
end
results{end+1, 1} = 'parity round3: per-attribute batched route matches enumeration at K = r';
results{end, 2} = all(abs(I(:) - Iref(:)) <= 1e-12 * max(abs(Iref(:))));

% --- B-16: the closed form refuses an inner [rel] unit ---
x = rp3Flat(14, 1.0, 2, true, false, 'full-image', rp3_P, 4, 1);
cx = mobius.closedFormAttrCentres(internal.prunedExpTens(x), 1);
Mok = mobius.closedFormAttrMatrixFrom(cx, cx, 'full-image');
cxBad = cx; cxBad.innerBlockSize = 2;
results{end+1, 1} = 'parity round3: closed form serves a flat bundle';
results{end, 2} = isequal(size(Mok), [1 1]) && isfinite(Mok);
results{end+1, 1} = 'parity round3: closed form refuses an inner-unit bundle';
results{end, 2} = throwsErrorWithId( ...
    @() mobius.closedFormAttrMatrixFrom(cxBad, cx, 'full-image'), ...
    'mobius:closedFormAttrMatrixFrom:innerUnit');

% --- D: the orphans are gone ---
rp3_root = fileparts(fileparts(mfilename('fullpath')));
rp3_gone = {fullfile(rp3_root, '+internal', 'warnRelPerAllImage.m'), ...
            fullfile(rp3_root, '+mobius', 'contract.m'), ...
            fullfile(rp3_root, '+mobius', 'innerProductDirectAbsSingleMultiset.m')};
results{end+1, 1} = 'parity round3: deleted and moved routines are gone from the shipped tree';
results{end, 2} = ~any(cellfun(@(f) exist(f, 'file') == 2, rp3_gone)) ...
    && exist(fullfile(rp3_root, 'tests', 'reference', '+reference', ...
                      'contract.m'), 'file') == 2;
rp3_src = fileread(fullfile(rp3_root, '+mobius', 'maPerAttrInnerMatrix.m'));
results{end+1, 1} = 'parity round3: maPerAttrInnerMatrix carries no safe/unsafe partition';
results{end, 2} = ~contains(rp3_src, 'localFillDirectEnumGroups') ...
    && ~contains(rp3_src, 'localPackNanTop') ...
    && ~contains(rp3_src, 'safe_x_mask');
rp3_src = fileread(fullfile(rp3_root, '+internal', 'predictOrbitCostMs.m'));
results{end+1, 1} = 'parity round3: predictOrbitCostMs carries no unused constants';
results{end, 2} = ~contains(rp3_src, 'GRID_OP') ...
    && ~contains(rp3_src, 'CENTRES_OP') && ~contains(rp3_src, 'REL_BASE');

% --- E: the retired cancellationThreshold keyword is rejected ---
% The keyword is gone from the parser, so it is left in varargin and
% trips the usual argument-count usage error rather than being accepted
% and ignored.
x = rp3Flat(21, 1.0, 2, false, false, 'full-image', rp3_P, 4, 1);
y = rp3Flat(22, 1.0, 2, false, false, 'full-image', rp3_P, 4, 1);
results{end+1, 1} = 'parity round3: cancellationThreshold keyword is rejected';
results{end, 2} = throwsErrorWithId( ...
    @() cosSimExpTens(x, y, 'cancellationThreshold', 1e-12, 'verbose', false), ...
    'cosSimExpTens:wrongArgCount');

clear x y z rep repCold repWarm xOut yOut selIn selIn2 orderedAny nestedAny ...
      dSmall dLarge mSmall mLarge Px Wx Py Wy I Iref ii jj cx cxBad Mok pZ

if standalone
    nPass = sum([results{:, 2}]);
    nFail = size(results, 1) - nPass;
    for ii = 1:size(results, 1)
        if results{ii, 2}
            fprintf('  PASS  %s\n', results{ii, 1});
        else
            fprintf('  FAIL  %s\n', results{ii, 1});
        end
    end
    fprintf('\n=== test_routing_parity_round3: %d passed, %d failed (of %d) ===\n\n', ...
        nPass, nFail, nPass + nFail);
    clear cleanupDefaults_rp3 rp3_warnCleanup
    if nFail > 0
        error('test_routing_parity_round3:failed', '%d test(s) failed.', nFail);
    end
end


function d = rp3Flat(seed, sigma, r, isRel, isPer, wrap, P, K, N)
    if nargin < 8; K = 5; end
    if nargin < 9; N = 2; end
    rng(seed, 'twister');
    p = sort(12.0 * rand(K, N), 1);
    if isPer
        period = P;
    else
        period = 0;
    end
    d = buildExpTens({p}, {[]}, sigma, r, isRel, isPer, period, ...
                     'wrap', {wrap}, 'verbose', false);
end


function tf = rp3RouteOf(densOut, route)
%RP3ROUTEOF  Whether the memo the call wrote on this operand carries the
%   given route's key (the key prefix names the route that ran).
    tf = isfield(densOut, 'selfIP') && isstruct(densOut.selfIP) ...
        && any(strncmp(densOut.selfIP.keys, [route '|'], numel(route) + 1));
end

File: python/tests/test_routing_parity_round3.py
"""Regressions from the routing-parity audit, round 3 (September 2026).

Mirror of ``tests/test_routing_parity_round3.m``. The round removed the
dead code the audit found and made ``explain_dispatch`` report the route
the call takes; each block pins one of those outcomes so the two
languages keep agreeing:

* ``explain_dispatch`` on a flat cosine builds the selector's inputs by
  the same function the call uses (the wrap vector, the per-attribute
  grid node counts, the memo flags), applies the empty-operand rule
  before the selector and the ordered-attribute rule after it --- so on
  a rel-per pair above the sigma/P threshold the report follows the
  declared wrap as the call does (it used to name the cost model's
  pick), and on an ordered pair it names Bulger's method under any
  ``method``;
* the eval cost model prices the spectral branch of the Möbius
  relative evaluator wherever that branch engages: the evaluator has no
  mode-grid size decline, so the model no longer prices one (D-6);
* the Möbius per-attribute matrix has one abs r >= 2 route: the
  safe/unsafe partition and its direct-enumeration fill are gone, and
  the batched route agrees with the enumerated reference at K = r;
* the tuple-centres closed form refuses an inner ``[rel]`` unit rather
  than carrying a block-Q form no caller can reach;
* the deleted names stay deleted (no re-export brings them back).
"""
import math
import warnings

import numpy as np
import pytest

import mpt
from mpt import build_exp_tens, cos_sim_exp_tens, explain_dispatch
from mpt._tensor import cosine as _cos
from mpt._tensor import dispatch as _disp
from mpt._tensor._mobius_inner import (_closed_form_attr_matrix_from,
                                       _ma_per_attr_inner_matrix)
from tests.references.mobius_ip_reference import inner_product_direct_abs


@pytest.fixture(autouse=True)
def _quiet():
    mpt.set_default(show_hints=False)
    warnings.simplefilter("ignore", UserWarning)
    yield
    warnings.resetwarnings()
    mpt.reset_defaults()


P = 12.0


def _flat(seed, sigma, r=2, *, is_rel=False, is_per=False,
          wrap='full-image', K=5, N=2, is_sym=True, weight=None):
    rng = np.random.default_rng(seed)
    p = np.sort(rng.uniform(0.0, P, size=(K, N)), axis=0)
    w = None if weight is None else [np.full((K, N), float(weight))]
    return build_exp_tens([p], w, [sigma], [r], [is_rel], [is_per],
                          [P if is_per else 0.0], [is_sym], wrap=[wrap],
                          verbose=False)


def _route_taken(monkeypatch):
    """Record which flat inner-product route ``cos_sim_exp_tens`` runs."""
    taken = []
    orbit = _cos._cos_sim_exp_tens_ma_orbit
    pairwise = _cos._cos_sim_exp_tens_ma_pairwise

    def _orbit(*a, **k):
        taken.append("mobius")
        return orbit(*a, **k)

    def _pairwise(*a, **k):
        taken.append("bulger")
        return pairwise(*a, **k)

    monkeypatch.setattr(_cos, "_cos_sim_exp_tens_ma_orbit", _orbit)
    monkeypatch.setattr(_cos, "_cos_sim_exp_tens_ma_pairwise", _pairwise)
    return taken


# ---------------------------------------------------------------------
# explain_dispatch reports the route the call takes
# ---------------------------------------------------------------------

@pytest.mark.parametrize("wrap, expected", [("full-image", "mobius"),
                                            ("single-image", "bulger")])
def test_explain_follows_the_wrap_rule_above_the_threshold(
        monkeypatch, wrap, expected):
    """r = 3, K = 4 vs 5, rel-per at sigma/P = 0.5: above the threshold
    the declared wrap decides without pricing. The report used to omit
    the wrap vector and so named the cost model's pick."""
    x = _flat(1, 0.5 * P, r=3, is_rel=True, is_per=True, K=4, wrap=wrap)
    y = _flat(2, 0.5 * P, r=3, is_rel=True, is_per=True, K=5, wrap=wrap)
    exp = explain_dispatch(x, y)
    assert exp.chosen == expected
    assert exp.decided_by == "structural rule"
    taken = _route_taken(monkeypatch)
    cos_sim_exp_tens(x, y, verbose=False)
    assert taken == [expected]
    # The measure line follows the same reading.
    assert ("transposition average" in exp.measure) == (expected == "mobius")


@pytest.mark.parametrize("method", ["auto", "mobius", "centres"])
def test_explain_applies_the_ordered_attribute_rule(monkeypatch, method):
    """An ordered ([sym]=0) attribute at r > 1 has no orbit: the call
    takes Bulger's method whatever ``method`` asked for, and so does the
    report."""
    x = _flat(3, 1.0, r=2, is_sym=False, K=5)
    y = _flat(4, 1.0, r=2, is_sym=False, K=6)
    exp = explain_dispatch(x, y, method=method)
    assert exp.chosen == "bulger"
    assert "ordered" in exp.decided_by
    taken = _route_taken(monkeypatch)
    cos_sim_exp_tens(x, y, method=method, verbose=False)
    assert taken == ["bulger"]


def test_explain_uses_the_memo_flags_the_call_uses():
    """After a call has memoised both self inner products, the report
    prices the cross matrix alone, as the call does: its Bulger price
    drops below the cold report's."""
    x = _flat(5, 1.0, r=3, is_rel=True, is_per=False, K=6, N=3)
    y = _flat(6, 1.0, r=3, is_rel=True, is_per=False, K=6, N=3)
    cold = explain_dispatch(x, y)
    cos_sim_exp_tens(x, y, method="bulger", verbose=False)
    warm = explain_dispatch(x, y)
    pw_cold = next(r.predicted_ms for r in cold.routes if r.name == "bulger")
    pw_warm = next(r.predicted_ms for r in warm.routes if r.name == "bulger")
    assert pw_warm < pw_cold


def test_explain_applies_the_empty_operand_rule():
    """An operand that prunes to no events returns 0.0 before any
    selector runs; the report says so and names no route."""
    x = _flat(7, 1.0, r=2, weight=0.0)
    y = _flat(8, 1.0, r=2)
    exp = explain_dispatch(x, y)
    assert exp.chosen is None
    assert "empty operand" in exp.decided_by
    assert cos_sim_exp_tens(x, y, verbose=False) == 0.0


def test_selector_inputs_are_shared_with_the_call():
    """The report and the call build the selector's inputs by one
    function, so they cannot drift: its wrap vector, node counts and
    memo flags are those of the densities."""
    x = _flat(9, 0.5 * P, r=3, is_rel=True, is_per=True, K=4,
              wrap="single-image")
    y = _flat(10, 0.5 * P, r=3, is_rel=True, is_per=True, K=5,
              wrap="single-image")
    kw, ordered_any, nested_any = _cos._flat_selector_inputs(
        x.pruned(), y.pruned(), normalize="cosine", truncation_sigmas=None)
    assert kw["wrap_vec"] == ["single-image"]
    assert kw["per_vec"] == [True]
    assert kw["nu_vec"][0] > 1 and not ordered_any and not nested_any
    assert kw["skip_xx"] is False and kw["skip_yy"] is False
    cos_sim_exp_tens(x, y, verbose=False)
    kw2, _o, _n = _cos._flat_selector_inputs(
        x.pruned(), y.pruned(), normalize="cosine", truncation_sigmas=None)
    assert kw2["skip_xx"] is True and kw2["skip_yy"] is True


# ---------------------------------------------------------------------
# D-6: the eval cost model prices the spectral branch that runs
# ---------------------------------------------------------------------

def test_eval_cost_model_has_no_mode_grid_decline():
    """A periodic r = 4 relative attribute at small sigma/P passes the
    evaluator's spectral gate; the model prices that branch (its cost
    is the K-free per-mode slope, so it does not grow with K as the
    node path's tabulation term does)."""
    sigma = 0.004 * P          # a mode grid that the old decline refused
    small = _flat(11, sigma, r=4, is_rel=True, is_per=True, K=16, N=1)
    large = _flat(12, sigma, r=4, is_rel=True, is_per=True, K=48, N=1)
    n_q = 256
    _c1, m_small = _disp._ma_eval_costs_ms(small, n_q)
    _c2, m_large = _disp._ma_eval_costs_ms(large, n_q)
    assert math.isfinite(m_small) and math.isfinite(m_large)
    # The r = 4 periodic K term is zero, so the spectral price is the
    # same for both value counts; the node path it used to fall back to
    # carries a K-proportional tabulation term.
    assert m_large == pytest.approx(m_small, rel=1e-12)


# ---------------------------------------------------------------------
# B-16: one abs r >= 2 route in the per-attribute matrix
# ---------------------------------------------------------------------

def test_per_attr_matrix_batched_route_agrees_with_enumeration_at_k_equal_r():
    """With every event at K_eff = r the former partition sent every
    pair to direct enumeration; the one batched route agrees with that
    enumeration on the value scale."""
    rng = np.random.default_rng(13)
    Px = np.sort(rng.uniform(0.0, 40.0, size=(3, 4)), axis=0)
    Wx = np.ones_like(Px)
    Py = np.sort(rng.uniform(0.0, 40.0, size=(3, 5)), axis=0)
    Wy = np.ones_like(Py)
    sigma, r = 3.0, 3
    # Compare exhaustively: at the 1e-12 accuracy floor the two drop
    # marginally different far-tail terms, which the alternating sum
    # amplifies on the near-zero cross entries.
    from mpt._defaults import accuracy_floor_context
    with accuracy_floor_context(1e-300):
        I = _ma_per_attr_inner_matrix(Px, Wx, Py, Wy, sigma, r, False, False,
                                      0.0, truncation_sigmas=float("inf"))
    ref = np.array([[inner_product_direct_abs(Px[:, i], Wx[:, i],
                                              Py[:, j], Wy[:, j],
                                              sigma, r, False, 0.0)
                     for j in range(5)] for i in range(4)])
    np.testing.assert_allclose(I, ref, rtol=0.0,
                               atol=1e-12 * float(np.max(np.abs(ref))))


def test_per_attr_matrix_has_no_partition_helpers():
    from mpt._tensor import _mobius_inner as _mi
    for name in ("_batched_direct_enum_abs", "_pack_nan_top"):
        assert not hasattr(_mi, name)
        assert not hasattr(mpt.tensor, name)


# ---------------------------------------------------------------------
# B-16: the closed form refuses an inner [rel] unit
# ---------------------------------------------------------------------

def test_closed_form_refuses_an_inner_unit_bundle():
    """A bundle whose inner block size is >= 2 cannot reach the closed
    form from any shipped plan; handed one directly, it refuses rather
    than computing a flat quadratic form over a block-diagonal metric."""
    x = _flat(14, 1.0, r=2, is_rel=True, K=4, N=1)
    from mpt._tensor._mobius_inner import _closed_form_attr_centres
    cx = list(_closed_form_attr_centres(x.pruned(), 0))
    ok = _closed_form_attr_matrix_from(tuple(cx), tuple(cx))
    assert ok.shape == (1, 1)
    cx[3] = 2                        # pretend an inner unit of size 2
    with pytest.raises(ValueError, match="inner \\[rel\\] unit"):
        _closed_form_attr_matrix_from(tuple(cx), tuple(cx))


# ---------------------------------------------------------------------
# D: the orphans are gone
# ---------------------------------------------------------------------

def test_deleted_names_stay_deleted():
    gone = {
        _cos: ("_rel_contract_cheaper", "_attr_value_range", "_ma_has_nan",
               "_orbit_inner_abs", "_orbit_inner_rel",
               "_inner_product_direct_abs", "_build_ordered_r_tuples"),
        _disp: ("_orbit_ips_impossible", "_estimate_centres_array_bytes",
                "_pw_per_entry_ms", "_PW_PER_ENTRY_MS_NONPER",
                "_PW_PER_ENTRY_MS_PER", "_warn_rel_per_all_image",
                "_format_time", "_PROBE_MIN_N_Q", "_PROBE_K_IP_TARGET",
                "_PROBE_TIME_CACHE", "_PROBE_IP_MOBIUS_DECISION_MARGIN",
                "_PRESCREEN_IP_DOMINANCE", "_ORBIT_IP_FIXED_OVERHEAD",
                "_CENTRES_PROBE_MEM_BUDGET"),
        mpt.tensor: ("_orbit_ips_impossible", "_estimate_centres_array_bytes",
                     "_PROBE_MIN_N_Q", "_PROBE_K_IP_TARGET",
                     "_PRESCREEN_IP_DOMINANCE", "_orbit_inner_abs",
                     "_orbit_inner_rel", "_ma_has_nan"),
    }
    for mod, names in gone.items():
        for name in names:
            assert not hasattr(mod, name), f"{mod.__name__}.{name}"
    from mpt._tensor import sweep as _sweep
    assert not hasattr(_sweep, "SweepNotEligible")
    from mpt._tensor import _nested_contraction as _nc
    for name in ("nested_ip", "make_quadrature", "_ip_absolute",
                 "_ip_rel_nonper", "_theta_truncation_L"):
        assert not hasattr(_nc, name)
    import importlib
    with pytest.raises(ImportError):
        importlib.import_module("mpt._tensor._centres_inner")
    # The hard budget kept its value under its honest name.
    assert _disp._DISPATCH_MEM_BUDGET == 4 * 1024 ** 3


# ---------------------------------------------------------------------
# E: the retired cancellation_threshold keyword is rejected
# ---------------------------------------------------------------------

def test_cancellation_threshold_keyword_is_rejected():
    """``cancellation_threshold`` is gone; passing it raises the normal
    unknown-keyword error rather than being accepted and ignored."""
    x = _flat(21, 1.0, r=2, K=4, N=1)
    y = _flat(22, 1.0, r=2, K=4, N=1)
    with pytest.raises(TypeError, match="cancellation_threshold"):
        cos_sim_exp_tens(x, y, cancellation_threshold=1e-12, verbose=False)

